packages/docusaurus/src/client/clientEntry.tsx

```diff
@@ -30,16 +30,18 @@ if (ExecutionEnvironment.canUseDOM) {
   // first-load experience.
   // For development, there is no existing markup so we had to render it.
   // We also preload async component to avoid first-load loading screen.
-  const renderMethod =
-    process.env.NODE_ENV === 'production' ? ReactDOM.hydrate : ReactDOM.render;
+  const container = document.getElementById('__docusaurus')!;
+  const renderMethod = (app: ReactChild) =>
+    process.env.NODE_ENV === 'production'
+      ? ReactDOM.hydrateRoot(container, app)
```

Comment:
We'll also need to update to renderToPipeableStream on the server
https://twitter.com/reactjs/status/1511755578716028928

Although it might not be strictly required to make React 18 work, we can handle that in another PR later

Reply:
I've switched to the `renderToPipeableStream` API, please take a look